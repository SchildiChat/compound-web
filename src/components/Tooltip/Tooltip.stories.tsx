/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { Placement as PlacementType } from "@floating-ui/react";
import { Tooltip as TooltipComponent } from "./Tooltip";
import { IconButton } from "../Button";
import UserIcon from "@vector-im/compound-design-tokens/icons/user-profile.svg";
import { Meta, StoryFn } from "@storybook/react";
import React, { FC, ReactNode } from "react";

export default {
  title: "Tooltip",
  component: TooltipComponent,
  tags: ["autodocs"],
  controls: {
    include: ["placement", "open", "label", "caption", "isTriggerInteractive"],
  },
  argTypes: {
    placement: {
      control: "inline-radio",
      options: ["top", "right", "left", "bottom"],
    },
    open: {
      control: "boolean",
    },
    isTriggerInteractive: {
      control: "boolean",
    },
    label: {
      control: "string",
    },
    caption: {
      control: "string",
    },
  },
  args: {
    placement: "left",
    label: "@bob:example.org",
    // needed, to prevent the tooltip to be in controlled mode
    onOpenChange: undefined,
    children: (
      <IconButton data-testid="testbutton">
        <UserIcon />
      </IconButton>
    ),
  },
  decorators: [
    (Story: StoryFn) => (
      <div style={{ padding: 100 }}>
        <Story />
      </div>
    ),
  ],
} as Meta<typeof TooltipComponent>;

interface LayoutProps {
  children: ReactNode;
}

const Layout: FC<LayoutProps> = ({ children }) => (
  <div
    style={{
      display: "flex",
      gap: "50px",
      flexDirection: "column",
      alignItems: "center",
    }}
  >
    {children}
  </div>
);

const TemplatePlacement: StoryFn<typeof TooltipComponent> = () => (
  <Layout>
    {(
      [
        "top",
        "top-start",
        "right",
        "right-end",
        "bottom",
        "bottom-end",
        "left",
        "left-start",
      ] as Array<PlacementType>
    ).map((placement) => (
      <TooltipComponent
        key={placement}
        open={true}
        placement={placement}
        label="@bob:example.org"
      >
        <IconButton>
          <UserIcon />
        </IconButton>
      </TooltipComponent>
    ))}
  </Layout>
);

export const Placement = TemplatePlacement.bind({});
Placement.args = {};

export const Default = {
  args: {
    // unset to test defaults
    placement: undefined,
  },
};

export const WithStringCaption = {
  args: {
    ...Default.args,
    label: "I can have a caption",
    caption: "My beautiful caption",
  },
};

export const WithComponentCaption = {
  args: {
    ...Default.args,
    label: "Copy",
    caption: (
      <div>
        <kbd>Ctrl</kbd> + <kbd>C</kbd>
      </div>
    ),
  },
};

export const ForcedOpen = {
  args: {
    ...Default.args,
    open: true,
    label: "I'm always open",
  },
};

export const ForcedClose = {
  args: {
    ...Default.args,
    open: false,
    label: "You can't see me",
    children: <span>No tooltip to see here</span>,
  },
};

export const InteractiveTrigger = {
  args: {
    ...Default.args,
    isTriggerInteractive: true,
    label: "Shown with delay",
  },
};

export const NonInteractiveTrigger = {
  args: {
    ...Default.args,
    isTriggerInteractive: false,
    label: "Shown without delay",
    children: (
      <IconButton data-testid="testbutton" disabled>
        <UserIcon />
      </IconButton>
    ),
  },
};
