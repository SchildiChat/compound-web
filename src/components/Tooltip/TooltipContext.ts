/*
 *
 * Copyright 2024 New Vector Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * /
 */

import { createContext, useContext } from "react";
import { useTooltip } from "./useTooltip";

type ContextType = ReturnType<typeof useTooltip> | null;
/**
 * The context for the Tooltip components.
 */
export const TooltipContext = createContext<ContextType>(null);

/**
 * Provides the context for the Tooltip components.
 */
export function useTooltipContext() {
  const context = useContext(TooltipContext);

  if (context == null) {
    throw new Error("Tooltip components must be wrapped in <Tooltip />");
  }

  return context;
}
