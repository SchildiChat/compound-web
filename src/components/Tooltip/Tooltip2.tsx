/*
 *
 * Copyright 2024 New Vector Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * /
 */

import { TooltipContext, useTooltipContext } from "./TooltipContext";
import {
  FloatingArrow,
  FloatingPortal,
  Placement,
  useMergeRefs,
} from "@floating-ui/react";
import React, {
  PropsWithChildren,
  Ref,
  JSX,
  isValidElement,
  cloneElement,
} from "react";

import classNames from "classnames";
import styles from "./Tooltip.module.css";
import { useTooltip } from "./useTooltip";

type UseTooltipParam = Parameters<typeof useTooltip>[0];

interface TooltipProps
  extends Omit<UseTooltipParam, "placement" | "isTriggerInteractive"> {
  /**
   * The placement of the component
   * @default "right"
   */
  placement?: Placement;
  isTriggerInteractive?: boolean;
  nonInteractiveTriggerTabIndex?: number;
  label: string;
  caption?: string;
}

export function Tooltip2({
  children,
  placement = "bottom",
  isTriggerInteractive = true,
  nonInteractiveTriggerTabIndex = 0,
  label,
  caption,
  ...props
}: PropsWithChildren<TooltipProps>) {
  const context = useTooltip({ placement, isTriggerInteractive, ...props });

  return (
    <TooltipContext.Provider value={context}>
      <TooltipAnchor>
        {isTriggerInteractive ? (
          children
        ) : (
          <span tabIndex={nonInteractiveTriggerTabIndex}>{children}</span>
        )}
      </TooltipAnchor>
      <TooltipContent>
        <span id={context.labelId}>{label}</span>
        {/* Forcing dark theme, so that we have the correct contrast when
              using the text color secondary on a solid dark background.
              This is temporary and should only remain until we figure out
              the approach to on-solid tokens */}
        {caption && (
          <span className={classNames(styles.caption, "cpd-theme-dark")}>
            {caption}
          </span>
        )}
      </TooltipContent>
    </TooltipContext.Provider>
  );
}

function TooltipContent({
  children,
}: Readonly<PropsWithChildren>): JSX.Element | null {
  const { context: floatingContext, arrowRef, ...rest } = useTooltipContext();

  if (!floatingContext.open) return null;

  return (
    <FloatingPortal>
      <div
        ref={rest.refs.setFloating}
        aria-labelledby={rest.labelId}
        aria-describedby={rest.captionId}
        style={rest.floatingStyles}
        {...rest.getFloatingProps()}
        className={styles.tooltip}
      >
        <FloatingArrow
          ref={arrowRef}
          context={floatingContext}
          // design absolute value
          width={10}
          height={6}
          className={styles.arrow}
        />
        {children}
      </div>
    </FloatingPortal>
  );
}

function TooltipAnchor({
  children,
  ...props
}: Readonly<PropsWithChildren>): JSX.Element {
  const context = useTooltipContext();

  // The children can have a ref and we don't want to discard it
  // Doing a dirty cast to get the optional ref
  const childrenRef = (children as unknown as { ref?: Ref<HTMLElement> })?.ref;
  const ref = useMergeRefs([context.refs.setReference, childrenRef]);

  if (!isValidElement(children)) {
    throw new Error("Tooltip anchor must be a single valid React element");
  }

  return cloneElement(
    children,
    context.getReferenceProps({
      ref,
      ...props,
      ...children.props,
      "data-state": context.open ? "open" : "closed",
    }),
  );
}
