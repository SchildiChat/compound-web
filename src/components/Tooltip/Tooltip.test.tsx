/*
Copyright 2023-2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { describe, it, expect, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import React, { act } from "react";

import * as stories from "./Tooltip.stories";
import { composeStories, composeStory } from "@storybook/react";
import userEvent from "@testing-library/user-event";
import { TooltipProvider } from "./TooltipProvider";
import { Tooltip } from "./Tooltip";
import { UserIcon } from "@vector-im/compound-design-tokens/assets/web/icons";

const {
  Default,
  WithStringCaption,
  WithComponentCaption,
  ForcedOpen,
  ForcedClose,
  ForcedDisabled,
  InteractiveTrigger,
  NonInteractiveTrigger,
  Descriptive,
} = composeStories(stories);

/**
 * Patches an element to always match :focus-visible whenever it's in focus.
 * JSDOM doesn't seem to support this selector on its own.
 */
function mockFocusVisible(e: Element): void {
  const originalMatches = e.matches.bind(e);
  vi.spyOn(e, "matches").mockImplementation(
    (selectors) =>
      originalMatches(selectors) ||
      (selectors === ":focus-visible" && e === document.activeElement),
  );
}

describe("Tooltip", () => {
  it("renders open by default", () => {
    render(<ForcedOpen />);
    // tooltip labels button and does not use role="tooltip"
    screen.getByRole("button", { name: "I'm always open" });
    expect(screen.queryByRole("tooltip")).toBe(null);
  });

  it("renders closed by default", () => {
    render(<ForcedClose />);
    // no tooltip, and no label either
    expect(screen.queryByRole("tooltip")).toBe(null);
    expect(screen.queryByRole("button", { name: /.*/ })).toBe(null);
  });

  it("renders disabled", () => {
    render(<ForcedDisabled />);
    // no tooltip, and no label either
    expect(screen.queryByRole("tooltip")).toBe(null);
    expect(screen.queryByRole("button", { name: /.*/ })).toBe(null);
  });

  it("renders default tooltip", async () => {
    render(<Default />);
    // tooltip labels button and does not use role="tooltip"
    screen.getByRole("button", { name: "@bob:example.org" });
    expect(screen.queryByRole("tooltip")).toBe(null);
  });

  it("opens tooltip on focus", async () => {
    const user = userEvent.setup();
    render(<InteractiveTrigger />);
    mockFocusVisible(screen.getByRole("link"));
    expect(screen.queryByRole("tooltip")).toBe(null);
    await user.tab();
    // trigger focused, tooltip shown
    expect(screen.getByRole("link")).toHaveFocus();
    screen.getByRole("tooltip");
  });

  it("opens tooltip on focus where trigger is non interactive", async () => {
    const user = userEvent.setup();
    render(<NonInteractiveTrigger />);
    mockFocusVisible(screen.getByText("Just some text").parentElement!);
    expect(screen.queryByRole("tooltip")).toBe(null);
    await user.tab();
    // trigger focused, tooltip shown
    expect(screen.getByText("Just some text").parentElement).toHaveFocus();
    screen.getByRole("tooltip");
  });

  it("opens tooltip on long press", async () => {
    vi.useFakeTimers();
    try {
      render(<InteractiveTrigger />);
      expect(screen.queryByRole("tooltip")).toBe(null);
      // Press
      fireEvent.touchStart(screen.getByRole("link"));
      expect(screen.queryByRole("tooltip")).toBe(null);
      // And hold
      await act(() => vi.advanceTimersByTimeAsync(1000));
      screen.getByRole("tooltip");
      // And release
      fireEvent.touchEnd(screen.getByRole("link"));
      // Tooltip should remain visible for some time
      screen.getByRole("tooltip");
      await act(() => vi.advanceTimersByTimeAsync(2000));
      expect(screen.queryByRole("tooltip")).toBe(null);
    } finally {
      vi.useRealTimers();
    }
  });

  it("overrides default tab index for non interactive triggers", async () => {
    const user = userEvent.setup();
    const Component = composeStory(
      {
        ...stories.NonInteractiveTrigger,
        args: {
          ...stories.NonInteractiveTrigger.args,
          nonInteractiveTriggerTabIndex: -1,
        },
      },
      stories.default,
    );
    render(<Component />);
    await user.tab();
    // trigger cannot be focused
    expect(screen.queryByRole("tooltip")).toBe(null);
  });

  it("renders with string caption", async () => {
    const user = userEvent.setup();
    render(<WithStringCaption />);
    await user.tab();
    // tooltip labels button and describes button with caption
    expect(
      screen.getByRole("button", { name: "I can have a caption" }),
    ).toHaveAccessibleDescription("My beautiful caption");
  });

  it("renders with component caption", async () => {
    const user = userEvent.setup();
    render(<WithComponentCaption />);
    await user.tab();
    // tooltip labels button and describes button with caption
    expect(
      screen.getByRole("button", { name: "Copy" }),
    ).toHaveAccessibleDescription("Ctrl + C");
  });

  it("renders a descriptive tooltip", async () => {
    render(<Descriptive />);
    // tooltip shown, but does not change the button's accessible name
    screen.getByRole("tooltip", { name: "Employer Identification Number" });
    expect(screen.queryByRole("button", { name: "EIN" })).toBe(null);
  });

  it("labels an image", async () => {
    render(
      <TooltipProvider>
        <Tooltip isTriggerInteractive={false} label="User profile">
          <UserIcon role="image" width={24} height={24} />
        </Tooltip>
      </TooltipProvider>,
    );
    screen.getByRole("image", { name: "User profile" });
  });
});
