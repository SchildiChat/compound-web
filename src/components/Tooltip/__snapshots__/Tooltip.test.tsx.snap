// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip > opens tooltip on focus 1`] = `
<div
  aria-describedby=":rg:"
  aria-labelledby=":rg:"
  class="_tooltip_9b26f8"
  id=":ri:"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; transform: translate(5px, 6px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_9b26f8"
    height="10"
    style="position: absolute; pointer-events: none; bottom: 100%; transform: rotate(180deg); left: -1px;"
    viewBox="0 0 10 10"
    width="10"
  >
    <path
      d="M0,0 H10 L5,6 Q5,6 5,6 Z"
      stroke="none"
    />
    <clippath
      id=":rl:"
    >
      <rect
        height="10"
        width="10"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <span
    id=":rg:"
  >
    Shown with delay
  </span>
</div>
`;

exports[`Tooltip > opens tooltip on focus where trigger is non interactive 1`] = `
<div>
  <div
    style="padding: 100px;"
  >
    <span
      tabindex="0"
    >
      <button
        aria-disabled="true"
        class="_icon-button_e3253e"
        data-testid="testbutton"
        role="button"
        style="--cpd-icon-button-size: 32px;"
        tabindex="0"
      >
        <div
          class="_indicator-icon_ab1fc6"
          style="--cpd-icon-button-size: 100%; --cpd-color-icon-tertiary: var(--cpd-color-icon-disabled);"
        >
          <svg
            class="cpd-icon"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.175 13.825C9.958 14.608 10.9 15 12 15s2.042-.392 2.825-1.175C15.608 13.042 16 12.1 16 11s-.392-2.042-1.175-2.825C14.042 7.392 13.1 7 12 7s-2.042.392-2.825 1.175C8.392 8.958 8 9.9 8 11s.392 2.042 1.175 2.825Zm4.237-1.412A1.926 1.926 0 0 1 12 13c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 10 11c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 12 9c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412Z"
            />
            <path
              d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10Zm-2 0a8 8 0 1 0-16 0 8 8 0 0 0 16 0Z"
            />
            <path
              d="M16.23 18.792a12.47 12.47 0 0 0-1.455-.455 11.6 11.6 0 0 0-5.55 0c-.487.12-.972.271-1.455.455a8.04 8.04 0 0 1-1.729-1.454c.89-.412 1.794-.729 2.709-.95A13.76 13.76 0 0 1 12 16c1.1 0 2.183.13 3.25.387a14.78 14.78 0 0 1 2.709.95 8.042 8.042 0 0 1-1.73 1.455Z"
            />
          </svg>
        </div>
      </button>
    </span>
  </div>
</div>
`;

exports[`Tooltip > opens tooltip on focus where trigger is non interactive 2`] = `
<div
  aria-describedby=":rm:"
  aria-labelledby=":rm:"
  class="_tooltip_9b26f8"
  id=":ro:"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; transform: translate(5px, 6px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_9b26f8"
    height="10"
    style="position: absolute; pointer-events: none; bottom: 100%; transform: rotate(180deg); left: -1px;"
    viewBox="0 0 10 10"
    width="10"
  >
    <path
      d="M0,0 H10 L5,6 Q5,6 5,6 Z"
      stroke="none"
    />
    <clippath
      id=":rr:"
    >
      <rect
        height="10"
        width="10"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <span
    id=":rm:"
  >
    Shown without delay
  </span>
</div>
`;

exports[`Tooltip > overrides default tab index for non interactive triggers 1`] = `
<div>
  <div
    style="padding: 100px;"
  >
    <span
      tabindex="-1"
    >
      <button
        aria-disabled="true"
        class="_icon-button_e3253e"
        data-testid="testbutton"
        role="button"
        style="--cpd-icon-button-size: 32px;"
        tabindex="0"
      >
        <div
          class="_indicator-icon_ab1fc6"
          style="--cpd-icon-button-size: 100%; --cpd-color-icon-tertiary: var(--cpd-color-icon-disabled);"
        >
          <svg
            class="cpd-icon"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.175 13.825C9.958 14.608 10.9 15 12 15s2.042-.392 2.825-1.175C15.608 13.042 16 12.1 16 11s-.392-2.042-1.175-2.825C14.042 7.392 13.1 7 12 7s-2.042.392-2.825 1.175C8.392 8.958 8 9.9 8 11s.392 2.042 1.175 2.825Zm4.237-1.412A1.926 1.926 0 0 1 12 13c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 10 11c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 12 9c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412Z"
            />
            <path
              d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10Zm-2 0a8 8 0 1 0-16 0 8 8 0 0 0 16 0Z"
            />
            <path
              d="M16.23 18.792a12.47 12.47 0 0 0-1.455-.455 11.6 11.6 0 0 0-5.55 0c-.487.12-.972.271-1.455.455a8.04 8.04 0 0 1-1.729-1.454c.89-.412 1.794-.729 2.709-.95A13.76 13.76 0 0 1 12 16c1.1 0 2.183.13 3.25.387a14.78 14.78 0 0 1 2.709.95 8.042 8.042 0 0 1-1.73 1.455Z"
            />
          </svg>
        </div>
      </button>
    </span>
  </div>
</div>
`;

exports[`Tooltip > renders closed by default 1`] = `
<DocumentFragment>
  <div
    style="padding: 100px;"
  >
    <span>
      No tooltip to see here
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tooltip > renders default tooltip 1`] = `
<div
  aria-describedby=":ra:"
  aria-labelledby=":ra:"
  class="_tooltip_9b26f8"
  id=":rc:"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; transform: translate(5px, 6px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_9b26f8"
    height="10"
    style="position: absolute; pointer-events: none; bottom: 100%; transform: rotate(180deg); left: -1px;"
    viewBox="0 0 10 10"
    width="10"
  >
    <path
      d="M0,0 H10 L5,6 Q5,6 5,6 Z"
      stroke="none"
    />
    <clippath
      id=":rf:"
    >
      <rect
        height="10"
        width="10"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <span
    id=":ra:"
  >
    @bob:example.org
  </span>
</div>
`;

exports[`Tooltip > renders open by default 1`] = `
<DocumentFragment>
  <div
    style="padding: 100px;"
  >
    <button
      aria-describedby=":r2:"
      class="_icon-button_e3253e"
      data-testid="testbutton"
      role="button"
      style="--cpd-icon-button-size: 32px;"
      tabindex="0"
    >
      <div
        class="_indicator-icon_ab1fc6"
        style="--cpd-icon-button-size: 100%;"
      >
        <svg
          class="cpd-icon"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.175 13.825C9.958 14.608 10.9 15 12 15s2.042-.392 2.825-1.175C15.608 13.042 16 12.1 16 11s-.392-2.042-1.175-2.825C14.042 7.392 13.1 7 12 7s-2.042.392-2.825 1.175C8.392 8.958 8 9.9 8 11s.392 2.042 1.175 2.825Zm4.237-1.412A1.926 1.926 0 0 1 12 13c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 10 11c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 12 9c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412Z"
          />
          <path
            d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10Zm-2 0a8 8 0 1 0-16 0 8 8 0 0 0 16 0Z"
          />
          <path
            d="M16.23 18.792a12.47 12.47 0 0 0-1.455-.455 11.6 11.6 0 0 0-5.55 0c-.487.12-.972.271-1.455.455a8.04 8.04 0 0 1-1.729-1.454c.89-.412 1.794-.729 2.709-.95A13.76 13.76 0 0 1 12 16c1.1 0 2.183.13 3.25.387a14.78 14.78 0 0 1 2.709.95 8.042 8.042 0 0 1-1.73 1.455Z"
          />
        </svg>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Tooltip > renders open by default 2`] = `
<div
  aria-describedby=":r0:"
  aria-labelledby=":r0:"
  class="_tooltip_9b26f8"
  id=":r2:"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_9b26f8"
    height="10"
    style="position: absolute; pointer-events: none; bottom: 100%; transform: rotate(180deg);"
    viewBox="0 0 10 10"
    width="10"
  >
    <path
      d="M0,0 H10 L5,6 Q5,6 5,6 Z"
      stroke="none"
    />
    <clippath
      id=":r5:"
    >
      <rect
        height="10"
        width="10"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <span
    id=":r0:"
  >
    I'm always open
  </span>
</div>
`;

exports[`Tooltip > renders with component caption 1`] = `
<DocumentFragment>
  <div
    style="padding: 100px;"
  >
    <button
      class="_icon-button_e3253e"
      data-testid="testbutton"
      role="button"
      style="--cpd-icon-button-size: 32px;"
      tabindex="0"
    >
      <div
        class="_indicator-icon_ab1fc6"
        style="--cpd-icon-button-size: 100%;"
      >
        <svg
          class="cpd-icon"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.175 13.825C9.958 14.608 10.9 15 12 15s2.042-.392 2.825-1.175C15.608 13.042 16 12.1 16 11s-.392-2.042-1.175-2.825C14.042 7.392 13.1 7 12 7s-2.042.392-2.825 1.175C8.392 8.958 8 9.9 8 11s.392 2.042 1.175 2.825Zm4.237-1.412A1.926 1.926 0 0 1 12 13c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 10 11c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 12 9c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412Z"
          />
          <path
            d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10Zm-2 0a8 8 0 1 0-16 0 8 8 0 0 0 16 0Z"
          />
          <path
            d="M16.23 18.792a12.47 12.47 0 0 0-1.455-.455 11.6 11.6 0 0 0-5.55 0c-.487.12-.972.271-1.455.455a8.04 8.04 0 0 1-1.729-1.454c.89-.412 1.794-.729 2.709-.95A13.76 13.76 0 0 1 12 16c1.1 0 2.183.13 3.25.387a14.78 14.78 0 0 1 2.709.95 8.042 8.042 0 0 1-1.73 1.455Z"
          />
        </svg>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Tooltip > renders with component caption 2`] = `
<div
  aria-describedby=":r17:"
  aria-labelledby=":r16:"
  class="_tooltip_9b26f8"
  id=":r18:"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; transform: translate(5px, 6px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_9b26f8"
    height="10"
    style="position: absolute; pointer-events: none; bottom: 100%; transform: rotate(180deg); left: -1px;"
    viewBox="0 0 10 10"
    width="10"
  >
    <path
      d="M0,0 H10 L5,6 Q5,6 5,6 Z"
      stroke="none"
    />
    <clippath
      id=":r1b:"
    >
      <rect
        height="10"
        width="10"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <span
    id=":r16:"
  >
    Copy
  </span>
  <div
    class="_caption_9b26f8 cpd-theme-dark"
    id=":r17:"
  >
    <kbd>
      Ctrl
    </kbd>
     + 
    <kbd>
      C
    </kbd>
  </div>
</div>
`;

exports[`Tooltip > renders with string caption 1`] = `
<DocumentFragment>
  <div
    style="padding: 100px;"
  >
    <button
      class="_icon-button_e3253e"
      data-testid="testbutton"
      role="button"
      style="--cpd-icon-button-size: 32px;"
      tabindex="0"
    >
      <div
        class="_indicator-icon_ab1fc6"
        style="--cpd-icon-button-size: 100%;"
      >
        <svg
          class="cpd-icon"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.175 13.825C9.958 14.608 10.9 15 12 15s2.042-.392 2.825-1.175C15.608 13.042 16 12.1 16 11s-.392-2.042-1.175-2.825C14.042 7.392 13.1 7 12 7s-2.042.392-2.825 1.175C8.392 8.958 8 9.9 8 11s.392 2.042 1.175 2.825Zm4.237-1.412A1.926 1.926 0 0 1 12 13c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 10 11c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 12 9c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412Z"
          />
          <path
            d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10Zm-2 0a8 8 0 1 0-16 0 8 8 0 0 0 16 0Z"
          />
          <path
            d="M16.23 18.792a12.47 12.47 0 0 0-1.455-.455 11.6 11.6 0 0 0-5.55 0c-.487.12-.972.271-1.455.455a8.04 8.04 0 0 1-1.729-1.454c.89-.412 1.794-.729 2.709-.95A13.76 13.76 0 0 1 12 16c1.1 0 2.183.13 3.25.387a14.78 14.78 0 0 1 2.709.95 8.042 8.042 0 0 1-1.73 1.455Z"
          />
        </svg>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Tooltip > renders with string caption 2`] = `
<div
  aria-describedby=":r11:"
  aria-labelledby=":r10:"
  class="_tooltip_9b26f8"
  id=":r12:"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; transform: translate(5px, 6px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_9b26f8"
    height="10"
    style="position: absolute; pointer-events: none; bottom: 100%; transform: rotate(180deg); left: -1px;"
    viewBox="0 0 10 10"
    width="10"
  >
    <path
      d="M0,0 H10 L5,6 Q5,6 5,6 Z"
      stroke="none"
    />
    <clippath
      id=":r15:"
    >
      <rect
        height="10"
        width="10"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <span
    id=":r10:"
  >
    I can have a caption
  </span>
  <span
    class="_caption_9b26f8 cpd-theme-dark"
    id=":r11:"
  >
    My beautiful caption
  </span>
</div>
`;
